### DISCUSSION TITLE
(ecs) Drain instances before replacing when updating launch template

### DISCUSSION THREAD

**Discussion by frjonsen** [aws/aws-cdk#24838] [Q&A]
I have an ECS cluster and an autoscaling group, which is added to the cluster using `addAsgCapacityProvider`.

The autoscaling group uses a Launch Template, which includes some CfInit.

I've noticed that if I change the launch template, for example changing instance type, it will immediately kill the old instances without draining them. The instances updates are at least rolling out one instance at a time, so of course I would prefer it if the tasks were moved to a new instance before the old one is killed.

I cannot figure out how to do this though. I've seen the properties `enableManagedScaling` and `enableManagedTerminationProtection`, but these already default to true and only seem to come into play during a scale-in.

Is there any way to achieve this, or do I just have to live with the system being unstable if I update the launch template?

**Comment by khushail** [comment]
Yes @frjonsen , there is a way to achieve this by using [lifecycleHook](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_autoscaling.LifecycleHook.html)` property of AutoScalingGroup to specify a Lambda function that will be invoked before the instance is terminated. In your lambda function, you can drain the tasks from the instance before its terminated.

```
{
  "lifecycleHook": {
    "lifecycleTransition": "autoscaling:EC2_INSTANCE_TERMINATING",
    "heartbeatTimeout": 300,
    "invocationType": "EC2_INSTANCE_INTERRUPTED",
    "target": {
      "arn": "arn:aws:lambda:us-east-1:123456789012:function:my-lambda-function"
    }
  }
}
```

Hope this helps !
